Fix part of #5343: Update Incorrect Link for the Oppia Android Code Coverage Wiki Page

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes Part of #5343 

### This PR includes

- Updated the incorrect link reference in the #5483 for the Oppia Android Code Coverage Page
  - Current Broken link: https://github.com/oppia/oppia-android-workflow/wiki/Oppia-Android-Code-Coverage
  - Correct link to wiki:  https://github.com/oppia/oppia-android/wiki/Oppia-Android-Code-Coverage
- Fixed an incorrect SKIP status that triggered even after 'Unit Tests - Bazel' failure causing it to miscalculate the pb file list as zero thereby posting a skip comment.
  - Solved by adding an additional condition for the evaluation job.
```
if: ${{ !cancelled() && needs.check_unit_tests_completed.result == 'success'}}
```
- The code coverage comment got triggered even after assignment changes due to the presence of 'assigned' in the on `pull_request_target` triggered, fixed it by removing the 'assigned' trigger.
- Utilized 'HtmlEscapers' to escape HTML characters from source code lines in CoverageReporter.kt as strings those already included html tags overlapped / conflicted with the report html templates.
- Added limitations section to the 'Oppia-Android-Code-Coverage' wiki page
- Filed all possible coverage gaps that exist right now based on the available pass cases.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

Rd4dev committed:
Fix part of #5343: Update Incorrect Link for the Oppia Android Code Coverage Wiki Page

Rd4dev commented:
@BenHenning, I incorrectly referenced the wrong link in PR #5483; updated to the correct link here, can you PTAL?

Oh, and also should we include other coverage gaps to the "[Limitations sections](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Code-Coverage#limitations-of-the-code-coverage-tool)"?

BenHenning commented:
Thanks @Rd4dev. I think adding other limitations sounds like a good idea.

Rd4dev commented:
Seems like I also had included assigned to the triggers,

```
on:
  pull_request_target:
    types: [assigned, opened, synchronize, reopened]
```

comment posted on assignment: https://github.com/oppia/oppia-android/pull/5511#issuecomment-2308493626
and its commenting on every assignment too :|, will add that fix too here ...

---

### Coverage Gaps

1. MathExpressionParser.kt -> As I don't understand the entire functionality just wanted to confirm if this is a missing parameter declaration causing the coverage miss or a jaCoCo coverage gap?

The lines are

```kotlin
  fun parseAlgebraicEquation(
    rawExpression: String,
    allowedVariables: List<String>,
    errorCheckingMode: ErrorCheckingMode = ErrorCheckingMode.ALL_ERRORS
  ): MathParsingResult<MathEquation> {
    println("Hitting...")
    return createAlgebraicParser(
      rawExpression, isPartOfEquation = true, allowedVariables, errorCheckingMode
    ).parseGenericEquationGrammar().map { it.stripParseInfo() }
  }
``` 

The test is:

```kotlin
  @Test
  fun testParseAlgebraicEquation_basicEquation_doesNotFail() {
    expectSuccessWhenParsingAlgebraicEquation("y = 2x + 3")
  }

    private fun parseAlgebraicEquation(expression: String): MathParsingResult<MathEquation> {
      return MathExpressionParser.parseAlgebraicEquation(
        expression, allowedVariables = listOf("x", "y", "z"), ALL_ERRORS
      )
    }
```

The report says:

![image](https://github.com/user-attachments/assets/5aaf014a-99ed-471f-bc3f-128370e509d7)

This was a similarly observed with methods: parseAlgebraicExpression() and parseNumericExpression().

#

2. **Abstract methods:**
Same as above I don't completely understand the inner working but the `TestCoroutineDispatcher.kt` has few abstract methods:

```kotlin
abstract fun hasPendingTasks(): Boolean

abstract fun getNextFutureTaskCompletionTimeMillis(timeMillis: Long): Long?

abstract fun runCurrent(
  timeout: Long = DEFAULT_TIMEOUT_SECONDS,
  timeoutUnit: TimeUnit = DEFAULT_TIMEOUT_UNIT
)
```

and these are implemented in _TestCorotineDispatcherRobolectricImpl.kt_, _TestCoroutineDispatchersRobolectricImpl.kt_, _TestCoroutineDispatchersEspressoImpl.kt_, etc.

But as per [jaCoCo documentation](https://www.jacoco.org/jacoco/trunk/doc/faq.html#:~:text=Abstract%20methods%20do%20not%20contain,non%2Ddefault%20methods%20in%20interfaces.):

>Abstract methods do not contain code, therefore code coverage cannot be evaluated. Indeed code coverage is recorded for subclasses implementing these methods

So, should the actual abstract methods in the TestCoroutineDispatcher.kt be uninstrumented as we only acquire coverage only from their respective test cases?

While most abstract methods are uninstrumented (highlighted as grey), the runCurrent shows as not covered (highlighted in red). Is this a coverage gap?

The html report:

![image](https://github.com/user-attachments/assets/a120b820-e72b-4674-8047-ea90ff38a32f)
![image](https://github.com/user-attachments/assets/6a890ef6-95c8-45a4-9e23-d914319ba18b)

#

3. **break:**

The finally block was reported as uncovered in `SurveyProgressController.kt`

```kotlin
    is ControllerMessage.FinishSurveySession -> {
      try {
        controllerState.completeSurveyImpl(message.callbackFlow)
      } finally {
        println("Finally...")
        // Ensure the actor ends since the session requires no further message processing.
        break
      }
    }
```

The report with break statement:

![image](https://github.com/user-attachments/assets/3b237780-3a5f-4547-a6bb-6e0ce2234db8)

But I think this will also align with the same exception gap, and the presence of break statement causes the partial coverage and commenting the break statement covers the finally block:

```kotlin
    } finally {
      println("Finally...")
      // Ensure the actor ends since the session requires no further message processing.
     // break
    }
```

The report without break statement: 

![image](https://github.com/user-attachments/assets/1cb6a801-be64-4223-8f58-8681c7b41a59)

If confirmed I will add it to the exception / termination statement #5506 issues.

#

4. **last curly brace:**

While this one is not observed in most cases, but with `CpuPerformanceSnapshotter.kt` the generated report highlights the last curly brace as uncovered:

The code:

```diff
    @OptIn(ObsoleteCoroutinesApi::class)
    private fun createCommandQueueActor(): SendChannel<CommandMessage> {
      return coroutineScope.actor(capacity = Channel.UNLIMITED) {
        for (message in channel) {
         ...
        }
-     }
    }
```

The generated report:

![image](https://github.com/user-attachments/assets/f31a3627-54a9-4727-b122-5b6dfedccaa1)

Similarly observed in CommandExecutorImpl.kt

The generated report:

![image](https://github.com/user-attachments/assets/b52b3e2b-bbe7-408a-9f9b-6a58d409ec1a)

also with TodoOpenCheck.kt

The generated report:

![image](https://github.com/user-attachments/assets/ce3e6112-90bc-4d86-ba2b-8bb4f3e853eb)

One related issue that can be found is: https://github.com/jacoco/jacoco/issues/1233

While I'm unsure if this is related to the test implementation or something that demands a configuration change or an actual coverage gap? If it is a coverage gap, then it will require further investigation on what specific cases they occur.

#

5. classes with just companion object

It is observed with `RepositoryFile.kt` that the class line was reported as uncovered, the generated report:

![image](https://github.com/user-attachments/assets/f553c08f-359a-412b-8fe7-92c7a6632b76)

and I doubt if they are related to :
- https://www.jacoco.org/jacoco/trunk/doc/faq.html#:~:text=If%20execution%20data%20is%20missing,from%20instrumentation%20or%20not%20executed.
- https://stackoverflow.com/questions/42348077/class-name-is-marked-as-not-covered-while-running-code-coverage-in-android-stu
- https://stackoverflow.com/questions/46477696/code-coverage-does-not-reach-class-declaration

same with TodoCollector.kt

![image](https://github.com/user-attachments/assets/b79132ef-c3c1-4a0b-9be4-246d5511563e)

---

- There seems to be a timeout issue with my local machine run so will fix than and rerun and try updating the results for GenerateMavenDependenciesList.kt, MavenDependenciesListCheck,  BazelClient.kt (if possible).
- And I am even afraid to attempt running RunCoverage.kt, so I’m going to pass on that
- Tried ComputeAffectedTestsTest, but I think I am skipping that too.
If mandatory I can try running coverage on the above files later.

I think I ran coverages on most if not all of the Passing cases (not on exempted overridden files as most are reported as 0 or very low and figuring out gaps would be not straightforward) and based on the other reports it can be seen that most of the uncovered lines are based on unhit else or other branch cases, overrides and I tried adding tests to see if they can be hit for few of them and they do seem to get covered. So, I think once all filed and (the above-mentioned gaps) to be filed issues are addressed, we can achieve move towards 100% code coverage. 

While I’m not sure if we’ll uncover more gaps once the source file incompatibility issues are resolved, that’s something we’ll only know once it’s done. Can you take a look at the above cases and direct if those can be filed as coverage gap or if any are uncovered with missing tests.

Once everything is filed, I will add them to the wiki limitations in one go.

---

### Other issues (Not coverage gaps)

HTML reports with source code containing html tags / other tags mess up the reports. For eg: With CoverageReporter.kt, LiTagHandler.kt, TodoCollector.kt reports, since it already have tags as strings within them, when the report is generated it mix matches the report structure. One possible way is to move these html strings as templates to a separate place and refer them from there so the source code report generation would still preserve the format.

eg: chopped - TodoCollector.kt

![image](https://github.com/user-attachments/assets/1f249044-8bad-4c8c-a9ff-cedb38a95c9b)

formatted CoverageReport.kt

![image](https://github.com/user-attachments/assets/4b9a07a0-2bf3-4281-a39b-a7eadec0621c)

LiTagHandler.kt

![image](https://github.com/user-attachments/assets/2767cf8d-1d24-4706-8d50-9f95105787eb)

Rd4dev committed:
Removed assignment trigger from the pull_request_target

BenHenning commented:
Thanks @Rd4dev. The code changes LGTM, but I'll need to process your last message a bit before I have any follow-up thoughts/suggestions.

BenHenning commented:
(1) is probably a real gap. It looks like the tests never verify the default parameter since they always provide one.

(2) is touching on several different things:

- We sometimes test abstract or interface classes as part of verifying their contracts (this is a case-by-case basis, and I don't think applies directly here).
- JaCoCo is making Java-specific assumptions (which are actually not correct anymore since interfaces in Java can have default methods). Its conclusion isn't really correct per modern Java or Kotlin.
- This is again, like (1), a case of default arguments seeming to not be verified so it's a real coverage gap on our end (assuming JaCoCo properly observes the coverage when it *is* tested).

(3) I'm not exactly sure if this is #5506 as-is, but I think it's reasonable to extend that issue to include this case. It definitely seems like a JaCoCo issue to me.

(4) Worth filing an issue for this case with the specific examples you've found. I suspect these are either coming from:

- A case where the code is simply not finishing in the test (in which case we need to add a properly configured test for each unit to verify that it "cleans up" correctly).
- Dead code being generated by the Kotlin compiler (similar to the Groovy problem mentioned in the JaCoCo issue that you linked).

Both will require investigation, and we need to rule out the first case before looking into the second one (which would require a bytecode analysis on the production code being instrumented).

(5) I suspect this is a case of the constructors not being called (since Kotlin generates public default constructors for classes) as we don't need to. If that's the case, we should just be turning the classes into public objects at the top-level if they don't need to be instances. This seems to be a coverage gap on our side, not a tooling issue.


-----

At a high level:
- I think skipping the tests you mentioned seems reasonable. The analysis you've performed so far looks very thorough and I'm happy to defer to your findings.
- It looks like one new issue should be filed (per (4)) and one issue updated (per (3)).
- The report generation can be fixed by escaping HTML tags (see https://stackoverflow.com/a/26572556/3689782 for a Guava-compatible way to do this without needing any additional dependencies). I think this is something that should be fixed, if possible, with a basic test in this PR.
- The wiki page should still be updated to include the coverage gaps once the issues above are filed.

Rd4dev committed:
Use HtmlEscapers to escape HTML symbols

Rd4dev commented:
@BenHenning, can you PTAL? and would it require any other verification on the Coverage Checks?

Ok, I found another case - https://github.com/oppia/oppia-android/pull/5522/checks, seems like evaluate runs irrespective of unit test failure and throwing "SKIP" status - https://github.com/oppia/oppia-android/pull/5522#issuecomment-2310730618, will also adress that over here.

Log:

- When Unit tests check fails now the Evaluate job gets skipped causing no misleading SKIP coverage report comment publication - https://github.com/oppia/oppia-android/actions/runs/10576551236/job/29304525595?pr=5511
- Added additional Unit test check pass to the Coverage Check Result job (to make sure coverage checks fails on unit test fail) - https://github.com/oppia/oppia-android/actions/runs/10576551236/job/29304525864?pr=5511
- Coverage SKIP to still get posted if no .kt files are changed - https://github.com/oppia/oppia-android/pull/5511#issuecomment-2312379269
- Comment being posted when everything goes well - https://github.com/oppia/oppia-android/pull/5511#issuecomment-2313034208

---

Thanks for the detailed information on each case. 
- I’ve filed a new issue as suggested #5523 and updated #5506 to include the relevant details. 
- I also applied the escape fix, and now the reports render right. For the testing I ended up using a direct conversion method
```kotlin
val unescapedOutputReportText = outputReportText
  .replace("&quot;", "\"")
  .replace("&amp;", "&")
  .replace("&lt;", "<")
  .replace("&gt;", ">")
```
not sure if there is anything to unescape (and escaping expected report caused issues with other tags).

Edit: <details><summary>Escaped CoverageReporter.html</summary>

![image](https://github.com/user-attachments/assets/3d03f579-f280-4ccf-ae24-d3f87325b95a)


</details>

- I've also added the newly filed issues to the limitations section of the wiki.

Rd4dev committed:
Fixes with evaluate code coverage report still running evaluation on unit test failure

Rd4dev committed:
Updating the limiations section to include the filed issues

Rd4dev committed:
Merge with develop branch

Rd4dev committed:
Reverting the CoverageReporter changes to make it a SKIP status

Rd4dev committed:
Updated the RunCoverageTest cases to take unescaped Html Report text

BenHenning commented on scripts/src/javatests/org/oppia/android/scripts/coverage/RunCoverageTest.kt:
```diff
@@ -317,7 +317,13 @@ class RunCoverageTest {
     val expectedResult = getExpectedHtmlText(kotlinFilePath)
 
-    assertThat(readHtmlReport(kotlinFilePath)).isEqualTo(expectedResult)
+    val unescapedHtmlReport = readHtmlReport(kotlinFilePath)
```
Could you maybe use Guava to unescape the HTML here & for other tests in the PR?

Rd4dev replied:
@BenHenning, Uhm... I'm not sure if Guava has methods to unescape the same?

[Guava Doc](https://guava.dev/releases/19.0/api/docs/com/google/common/escape/Escaper.html#escape(java.lang.String):~:text=Function%3CString,given%20literal%20string.)  |  [stackoverflow thread](https://stackoverflow.com/a/34091458)

or does it? or is there any other better way to do it than direct conversion.

BenHenning replied:
Good call, I suppose this approach seems reasonable then. I'd rather not introduce another dependency if we can avoid it, even for tests.

BenHenning reviewed (changes requested):
Thanks @Rd4dev! Generally LGTM, but had one comment--PTAL.

BenHenning commented:
@Rd4dev can you also please make sure the PR description is updated to include all of the new changes (since there have been a few iterations on this PR?).

BenHenning commented:
Hmm also I think the coverage report comment is reposting after PR event changes like assignees--are its triggers correct?

Rd4dev commented:
> Hmm also I think the coverage report comment is reposting after PR event changes like assignees--are its triggers correct?

Yeah, that shouldn't be triggering and spotted it earlier https://github.com/oppia/oppia-android/pull/5511#issuecomment-2308497331 and included a fix for that too here - [PR Changes](https://github.com/oppia/oppia-android/pull/5511/files#diff-8666350cec48d3bb6effe6965f6d25893c7729dfecf890f6a29ace6411b9c7d0R10)

Ah, I’ve updated the PR description with the details—hopefully, they’ll stand out better now and won’t get lost in the mix.

@BenHenning, can you please confirm on this https://github.com/oppia/oppia-android/pull/5511#discussion_r1733655294 or other approaches that could be followed. 

#

Edit: 
Seems like [Apache Commons](https://commons.apache.org/proper/commons-text/javadocs/api-release/org/apache/commons/text/StringEscapeUtils.html#escapeHtml4-java.lang.String-) does have unescape functionalities, but I think that will require additional dependency?

BenHenning reviewed (approved):
Ah, I forgot about the fact that the old version of the comment workflow is running since it's pull_request_target--thanks for pointing out that it should now be fixed @Rd4dev.

Since there's nothing left to resolve here, approving and merging. Thanks!
